RBasic-05-Arrays-Matrices/lesson.src.yaml

```diff
@@ -0,0 +1,323 @@
+  CorrectAnswer: x[,,1] + x[,,2]
+  AnswerTests: omnitest('x[,,1] + x[,,2]', x[,,1] + x[,,2])
+
+- Class : text
+  Output: 這裡相加的結果，就是把`x[,,1]`和`x[,,2]`對應的位置做相加。同學可以比對前面的輸出結果。
+
+- Class : cmd_question
+  Output: 而當維度不相同的時候，R 會自動重複比較短的那邊。當一邊是矩陣，一邊是單值的時候，很簡單。
+    請同學試試看：`x[,,1] + 1`
+  CorrectAnswer: x[,,1] + 1
+  AnswerTests: omnitest('x[,,1] + 1', x[,,1] + 1)
+  
+- Class : mult_question
+  Output: 但是當兩邊都是矩陣或陣列的時候，長度不一樣時，就不容易了。這裡我們先複習一下，陣列的值得排列方式。
+    如果`y <- array(1:8, c(2, 2, 2))`，那`y[1,1,1]`是1 、`y[2,1,1]`是2 。R 會把`1:8`的值依照第一個維度（列）、第二個維度（行）
+    到第三個維度，一個一個填入。以此類推，所以`y[1,2,1]`的值會是？
+  AnswerChoices: 3;4;5;6;7
+  CorrectAnswer: 3
+  AnswerTests: omnitest(correctVal = '3')
+
+- Class : cmd_question
+  Output: 當陣列和向量相加的時候，R 會一直重複向量，到它的長度和陣列需要的長度相符合。舉例來說，一個3 乘2 乘2 的陣列需要12個值。
+    接著，R 會把這個向量轉換成和陣列一樣的維度，然後在對應的位置做相加。請同學試試看：`matrix(1:4, 2, 2) + 1:2`
+  CorrectAnswer: matrix(1:4, 2, 2) + 1:2
+  AnswerTests: omnitest('matrix(1:4, 2, 2) + 1:2', matrix(c(1,4,4,6), 2, 2))
+
+- Class : cmd_question
+  Output: 另外我們提一個重要的事情，是關於上一個課程中有提到屬性的概念。現在的變數x 也是個R 物件，也有屬性。請同學查一下它的屬性。
+  CorrectAnswer: attributes(x)
+  AnswerTests: omnitest('attributes(x)', attributes(x))
+  Hint: 請用`attributes`這個函數。
+
+- Class : cmd_question
+  Output: 同學有沒有看到名稱叫做dim 的一個整數向量呢？這個dim 屬性非常重要，所有的矩陣和陣列，就是一般的向量加上
+    dim 這個屬性而已。R 也提供了`dim` 這個函數讓使用者可以存取dim 屬性。舉例來說，`dim(x)`會印出x 的維度。而我們
+    可以透過`dim(x) <- NULL`來移除x 的dim 屬性。請同學試試看`dim(x) <- NULL`
+  CorrectAnswer: dim(x) <- NULL
+  AnswerTests: omnitest('dim(x) <- NULL', NULL)
+
+- Class : cmd_question
+  Output: 請同學印出x 來看看。
+  CorrectAnswer: x
+  AnswerTests: omnitest('x', x)
+
+- Class : text
+  Output: 是不是就變成向量了呢？
+  
+- Class : text
+  Output: R 在陣列和向量的向量式運算，也可以回到兩個向量的運算。差別只是在它們多了維度的屬性，所以當維度差異太大的時候，
+    R 會認為向量式運算無效。所以一般來說，我們只會拿單值或向量去和陣列做運算，或是維度相同的矩陣或陣列做運算。
+
+- Class : text
+  Output: 基本上，`+`、`-`、`*`和`/`都會使用向量式運算。
+
+- Class : cmd_question
+  Output: 在R 中，使用`cbind`和`rbind`則可以合併兩個矩陣。舉例來說，`cbind(matrix(1:4, 2, 2), matrix(1:4, 2, 2))`會把兩個矩陣的行合併，
+    所以會變成2 乘 4的矩陣。請同學試試看。
+  CorrectAnswer: cbind(matrix(1:4, 2, 2), matrix(1:4, 2, 2))
+  AnswerTests: omnitest('cbind(matrix(1:4, 2, 2), matrix(1:4, 2, 2))')
+
+- Class : cmd_question
+  Output: 使用`rbind(matrix(1:4, 2, 2), matrix(1:4, 2, 2))`則會把兩個矩陣的列合併，
+    所以會變成4 乘 2的矩陣。請同學試試看。
+  CorrectAnswer: rbind(matrix(1:4, 2, 2), matrix(1:4, 2, 2))
+  AnswerTests: omnitest('rbind(matrix(1:4, 2, 2), matrix(1:4, 2, 2))')
+
+- Class : text
+  Output: 最後，我們介紹一些和線性代數相關的運算子和函數。如果對線性代數不熟悉的同學，可以在看過反矩陣之後
+    跳過（使用`skip()`）後面的課程。
+
+- Class : cmd_question
+  Output: 在R 中兩個矩陣要作矩陣乘法，就是使用`%*%`這個運算符號。請同學試試看：`matrix(1:6, 2, 3) %*% matrix(3:8, 3, 2)`
+  CorrectAnswer: matrix(1:6, 2, 3) %*% matrix(3:8, 3, 2)
+  AnswerTests: omnitest('matrix(1:6, 2, 3) %*% matrix(3:8, 3, 2)', matrix(1:6, 2, 3) %*% matrix(3:8, 3, 2))
+
+- Class : cmd_question
+  Output: 另外當兩個向量使用`%*%`做運算，會得到他們的內積。舉例來說：`1:3 %*% 1:3`會得到什麼呢？請同學試試看。
+  CorrectAnswer: 1:3 %*% 1:3
+  AnswerTests: omnitest('1:3 %*% 1:3', matrix(14, 1, 1))
```

Comment:
這邊也是比較奇怪的，

```r
| 另外當兩個向量使用`%*%`做運算，會得到他們的內積。舉例來說：`1:3 %*% 1:3`會得到什麼呢？請同學試試看。
> 1:3 %*% 1:3
     [,1]
[1,]   14

| One more time. You can do it! Or, type info() for more options.
| 1:3 %*% 1:3

> 1:3 %*% 1:3
     [,1]
[1,]   14

| Keep trying! Or, type info() for more options.
| 1:3 %*% 1:3

> 14
[1] 14

| That's not the expression I expected but it works.
| I've executed the correct expression in case the result is needed in an upcoming question.
```